bin/level/app.rs

```diff
@@ -165,6 +213,36 @@ impl<R: gfx::Resources> Application<R> for LevelView<R> {
                 let rot = cgmath::Quaternion::from_angle_x(cgmath::Rad(1.0 * delta * dir));
                 self.cam.rot = self.cam.rot * rot;
             }
+            Input::DepQuant {dir} => {
+                let vec = cgmath::Vector3::unit_z();
+                self.cam.loc += 1000.0 * delta * dir * vec.normalize();
+                self.input = Input::Empty;
+            }
+            Input::PlaneQuant {dir} => {
+                let cgmath::Vector2{x, y} = dir;
+
+                let vec_x = self.cam.rot * cgmath::Vector3::unit_x() * -x;
+                let vec_y = self.cam.rot * cgmath::Vector3::unit_y() * y;
+
+                let mut vec = vec_x + vec_y;
+
+                let norm1 = vec.magnitude();
+                let x1 = vec.x;
+                let y1 = vec.y;
+                vec.z = 0.0;
+                let norm = vec.magnitude();
+                vec = cgmath::vec3(x1 * norm1/norm, y1 * norm1/norm, 0.0);
+
+                self.cam.loc += self.cam.loc.z * 0.2  * delta * vec;
+                self.input = Input::Empty;
+            }
+            Input::RotQuant {dir} => {
+                let cgmath::Vector2{x, y} = dir;
```

Comment:
this doesn't help much here, you can either pattern match it earlier as `Input::RotQuant { dir: cgmath::Vector2 { x, y }} => {...}` or just use `dir.x` and `dir.y` here

Reply:
Yes, nested pattern matching! Didn't think about that. May be it's too verbose, i think i'll just switch to `dir.x` in this case.